File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    const hash = {
        ")": "(",
        "}": "{",
        "]": "[",
    }

    for (let char of s) {
        if (hash[char]) {
            if (stack.length === 0) return false;
            if (stack.pop() !== hash[char]) return false;
        } else {
            stack.push(char);
        }
    }

    return stack.length === 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        unordered_map<char, char> hash = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        for (int i = 0; i < s.size(); i++) {
            char curr = s[i];

            if (hash.find(curr) != hash.end()) {
                if (!st.empty() && st.top() == hash[curr]) {
                    st.pop();
                } else {
                    return false;
                }
            } else {
                st.push(curr);
            }
        }

        return st.empty();
    }
};